File: packages/sdk-go/conformance/manifest_test.go
package conformance

import (
	"encoding/json"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"testing"
)

func TestManifest(t *testing.T) {
	manifest := LoadManifest(t)
	if manifest.Version != 1 {
		t.Fatalf("manifest version = %d, want 1", manifest.Version)
	}

	expectedSuites := []FixtureSuite{
		{ID: "http/hmac-requests", Kind: "hmacRequest", Cases: []string{"private-post-with-body", "private-post-empty-body", "private-post-wide-integer", "websocket-upgrade-nonce"}},
		{ID: "http/unsigned-requests", Kind: "unsignedRequest", Cases: []string{"market-data-ticker", "market-data-order-book-limits", "prediction-markets-list-events-filters"}},
		{ID: "http/errors", Kind: "errorMapping", Cases: []string{"invalid-nonce-400", "missing-role-403-result-envelope", "missing-role-403-error-envelope", "terms-required-400-must-accept", "terms-required-400-accept-terms-required", "insufficient-funds-406", "rate-limited-429-retry-after", "order-not-found-404", "market-closed-400", "server-error-500-unstructured", "not-found-404-empty-body"}},
		{ID: "json", Kind: "jsonDecoding", Cases: []string{"wide-integer-unsafe", "wide-integer-safe", "decimal-string-exact", "decimal-number-exact"}},
		{ID: "websocket/subscriptions", Kind: "wsSubscription", Cases: []string{"public-trades", "public-trades-untrimmed-symbol", "public-book-ticker", "public-depth-diff", "public-depth-diff-interval", "public-partial-depth-20", "public-contract-status", "private-orders-session", "private-balances-interval"}},
		{ID: "websocket/events", Kind: "wsEvent", Cases: []string{"trade", "depth-unsafe-last-update-id", "order-update", "balance-update"}},
	}
	if len(manifest.Suites) != len(expectedSuites) {
		t.Fatalf("manifest suite count = %d, want %d", len(manifest.Suites), len(expectedSuites))
	}
	for i, want := range expectedSuites {
		got := manifest.Suites[i]
		if got.ID != want.ID || got.Kind != want.Kind {
			t.Errorf("suite %d = (%q, %q), want (%q, %q)", i, got.ID, got.Kind, want.ID, want.Kind)
		}
		if !equalStrings(got.Cases, want.Cases) {
			t.Errorf("suite %s cases = %v, want %v", got.ID, got.Cases, want.Cases)
		}
	}

	wantExceptions := []string{
		"rest-nonce-json-type",
		"rest-payload-key-order",
		"ws-subscription-id-scope",
		"ws-inbound-symbol-case",
		"int64-static-typing",
		"http-406-insufficient-funds",
		"rest-query-signing",
		"candle-timeframe-spelling",
	}
	if !equalStringSet(manifest.Exceptions, wantExceptions) {
		t.Fatalf("manifest exception ids = %v, want exact set %v", manifest.Exceptions, wantExceptions)
	}

	root := ConformanceRoot(t)
	conformanceDir := filepath.Join(root, "conformance")
	listed := make(map[string]struct{})
	referencedExceptions := make(map[string]struct{})
	for _, suite := range manifest.Suites {
		for _, caseID := range suite.Cases {
			rel := filepath.ToSlash(filepath.Join(suite.ID, caseID+".json"))
			if _, ok := listed[rel]; ok {
				t.Fatalf("manifest lists duplicate case %s", rel)
			}
			listed[rel] = struct{}{}
			path := filepath.Join(conformanceDir, filepath.FromSlash(rel))
			if _, err := os.Stat(path); err != nil {
				t.Errorf("manifest case %s is missing: %v", rel, err)
				continue
			}
			var envelope struct {
				Exceptions []string `json:"exceptions"`
			}
			data, err := os.ReadFile(path)
			if err != nil {
				t.Errorf("reading listed case %s: %v", rel, err)
				continue
			}
			if err := json.Unmarshal(data, &envelope); err != nil {
				t.Errorf("decoding listed case %s: %v", rel, err)
				continue
			}
			for _, id := range envelope.Exceptions {
				referencedExceptions[id] = struct{}{}
			}
		}
	}

	diskCases := make(map[string]struct{})
	if err := filepath.WalkDir(conformanceDir, func(path string, entry fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if entry.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(conformanceDir, path)
		if err != nil {
			return err
		}
		if filepath.ToSlash(rel) == "manifest.json" || filepath.Ext(path) != ".json" {
			return nil
		}
		diskCases[filepath.ToSlash(rel)] = struct{}{}
		return nil
	}); err != nil {
		t.Fatalf("walking conformance fixtures: %v", err)
	}
	for rel := range listed {
		if _, ok := diskCases[rel]; !ok {
			t.Errorf("manifest-listed case %s is not present on disk", rel)
		}
	}
	for rel := range diskCases {
		if _, ok := listed[rel]; !ok {
			t.Errorf("orphaned fixture file %s is not listed in manifest", rel)
		}
	}
	declaredExceptions := make(map[string]struct{}, len(manifest.Exceptions))
	for _, id := range manifest.Exceptions {
		declaredExceptions[id] = struct{}{}
	}
	for id := range referencedExceptions {
		if _, ok := declaredExceptions[id]; !ok {
			t.Errorf("case references undeclared exception %q", id)
		}
	}
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func equalStringSet(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	left := append([]string(nil), a...)
	right := append([]string(nil), b...)
	sort.Strings(left)
	sort.Strings(right)
	for i := range left {
		if left[i] != right[i] || left[i] == "" {
			return false
		}
	}
	return true
}

File: packages/sdk-typescript/src/tests/conformance/manifest.test.ts
import assert from "node:assert/strict";
import { readdir } from "node:fs/promises";
import { join, relative } from "node:path";
import { test } from "node:test";

import {
  conformanceRoot,
  loadCase,
  loadExceptions,
  loadManifest,
} from "./support/fixtures.js";

const EXPECTED_SUITES: Readonly<Record<string, readonly string[]>> = {
  "http/hmac-requests": [
    "private-post-with-body",
    "private-post-empty-body",
    "private-post-wide-integer",
    "websocket-upgrade-nonce",
  ],
  "http/unsigned-requests": [
    "market-data-ticker",
    "market-data-order-book-limits",
    "prediction-markets-list-events-filters",
  ],
  "http/errors": [
    "invalid-nonce-400",
    "missing-role-403-result-envelope",
    "missing-role-403-error-envelope",
    "terms-required-400-must-accept",
    "terms-required-400-accept-terms-required",
    "insufficient-funds-406",
    "rate-limited-429-retry-after",
    "order-not-found-404",
    "market-closed-400",
    "server-error-500-unstructured",
    "not-found-404-empty-body",
  ],
  json: [
    "wide-integer-unsafe",
    "wide-integer-safe",
    "decimal-string-exact",
    "decimal-number-exact",
  ],
  "websocket/subscriptions": [
    "public-trades",
    "public-trades-untrimmed-symbol",
    "public-book-ticker",
    "public-depth-diff",
    "public-depth-diff-interval",
    "public-partial-depth-20",
    "public-contract-status",
    "private-orders-session",
    "private-balances-interval",
  ],
  "websocket/events": [
    "trade",
    "depth-unsafe-last-update-id",
    "order-update",
    "balance-update",
  ],
};

const EXPECTED_KINDS: Readonly<Record<string, string>> = {
  "http/hmac-requests": "hmacRequest",
  "http/unsigned-requests": "unsignedRequest",
  "http/errors": "errorMapping",
  json: "jsonDecoding",
  "websocket/subscriptions": "wsSubscription",
  "websocket/events": "wsEvent",
};

const EXPECTED_EXCEPTIONS = [
  "rest-nonce-json-type",
  "rest-payload-key-order",
  "ws-subscription-id-scope",
  "ws-inbound-symbol-case",
  "int64-static-typing",
  "http-406-insufficient-funds",
  "rest-query-signing",
  "candle-timeframe-spelling",
] as const;

async function fixtureJsonFiles(root: string): Promise<string[]> {
  const files: string[] = [];
  async function visit(directory: string): Promise<void> {
    for (const entry of await readdir(directory, { withFileTypes: true })) {
      const path = join(directory, entry.name);
      if (entry.isDirectory()) {
        await visit(path);
      } else if (entry.isFile() && entry.name.endsWith(".json") && relative(root, path) !== "manifest.json") {
        files.push(relative(root, path).replace(/\.json$/, ""));
      }
    }
  }
  await visit(root);
  return files;
}

function sorted(values: Iterable<string>): string[] {
  return [...values].sort();
}

test("conformance manifest covers every fixture and declared exception", async () => {
  const manifest = await loadManifest();
  assert.equal(manifest.version, 1);
  assert.equal(manifest.suites.length, Object.keys(EXPECTED_SUITES).length);
  assert.deepEqual(sorted(manifest.suites.map((suite) => suite.id)), sorted(Object.keys(EXPECTED_SUITES)));

  const listedFiles = new Set<string>();
  const referencedExceptions = new Map<string, string[]>();
  for (const suite of manifest.suites) {
    assert.equal(suite.kind, EXPECTED_KINDS[suite.id]);
    assert.deepEqual(suite.cases, EXPECTED_SUITES[suite.id]);
    for (const caseId of suite.cases) {
      const relativePath = `${suite.id}/${caseId}`;
      listedFiles.add(relativePath);
      const value = await loadCase(suite.id, caseId);
      assert.equal(value.id, relativePath);
      assert.equal(value.kind, suite.kind);
      for (const exception of value.exceptions ?? []) {
        const ids = referencedExceptions.get(exception) ?? [];
        ids.push(relativePath);
        referencedExceptions.set(exception, ids);
      }
    }
  }

  const diskFiles = new Set(await fixtureJsonFiles(conformanceRoot()));
  assert.deepEqual(sorted(listedFiles), sorted(diskFiles), "manifest and fixture files must agree in both directions");

  assert.deepEqual(manifest.exceptions, EXPECTED_EXCEPTIONS);
  const exceptions = await loadExceptions();
  assert.deepEqual(sorted(exceptions.map((exception) => exception.id)), sorted(manifest.exceptions));
  const declaredExceptionIds = new Set(exceptions.map((exception) => exception.id));
  for (const exceptionId of referencedExceptions.keys()) {
    assert.ok(declaredExceptionIds.has(exceptionId), `case references undeclared exception ${exceptionId}`);
  }

  const declaredCases = new Map<string, string[]>();
  for (const exception of exceptions) {
    for (const caseId of exception.cases ?? []) {
      const cases = declaredCases.get(caseId) ?? [];
      cases.push(exception.id);
      declaredCases.set(caseId, cases);
    }
  }
  const observedCases = new Map<string, string[]>();
  for (const [exceptionId, cases] of referencedExceptions) {
    for (const caseId of cases) {
      const exceptionIds = observedCases.get(caseId) ?? [];
      exceptionIds.push(exceptionId);
      observedCases.set(caseId, exceptionIds);
    }
  }
  assert.deepEqual(
    sorted([...declaredCases.entries()].map(([caseId, ids]) => `${caseId}:${sorted(ids).join(",")}`)),
    sorted([...observedCases.entries()].map(([caseId, ids]) => `${caseId}:${sorted(ids).join(",")}`)),
    "overlay exception cases must match fixture references",
  );
});
